File: ClassLibrary1/screenshot_daemon.py
#!/usr/bin/env python3
"""
截图守护进程 - 运行在 Android 模拟器内部
TCP 服务，监听截图请求，返回 screencap raw 数据（跳过 PNG 编码，~100ms）
启动: python3 screenshot_daemon.py [端口号]
默认端口: 19000
协议:
  客户端发送: b"shot"
  服务端返回: b"<width> <height>\n" + raw RGBA 像素数据 (w*h*4 字节)
"""
import socket
import subprocess
import sys
import struct
import os

PORT = int(sys.argv[1]) if len(sys.argv) > 1 else 19000
SCREENCAP = "/system/bin/screencap"

def capture_screen():
    """调用 screencap 获取 raw 数据，返回 (width, height, rgba_bytes)"""
    raw = subprocess.check_output([SCREENCAP])
    # raw 格式: 12 字节头 (width:i32le, height:i32le, format:i32le) + 像素数据
    if len(raw) < 12:
        return None
    w, h = struct.unpack_from('<II', raw, 0)
    pixels = raw[12:]
    return w, h, pixels

def handle_client(conn):
    """处理单个客户端请求"""
    try:
        data = conn.recv(64)
        if data.strip() == b"shot":
            result = capture_screen()
            if result is None:
                conn.sendall(b"ERR\n")
                return
            w, h, pixels = result
            # 发送文本头 + 二进制像素数据
            header = "{} {}\n".format(w, h).encode()
            conn.sendall(header)
            conn.sendall(pixels)
        elif data.strip() == b"ping":
            conn.sendall(b"pong\n")
    except Exception as e:
        try:
            conn.sendall(b"ERR\n")
        except:
            pass

def main():
    srv = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    srv.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    srv.bind(("127.0.0.1", PORT))
    srv.listen(4)
    print("Screenshot daemon listening on port {}".format(PORT), flush=True)
    while True:
        conn, addr = srv.accept()
        try:
            handle_client(conn)
        finally:
            conn.close()

if __name__ == "__main__":
    main()

File: ClassLibrary1/screenshot_daemon.c
/* screenshot_daemon.c - runs inside Android emulator
 * Compile: gcc -o screenshot_daemon screenshot_daemon.c
 * Usage:   screenshot_daemon [port]
 * Protocol: client sends "shot" -> receives raw screencap output (12B header + RGBA pixels)
 *           client sends "ping" -> receives "pong\n"
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <sys/wait.h>
#include <signal.h>
#include <errno.h>

static int send_screenshot(int client_fd) {
    int pipefd[2];
    if (pipe(pipefd) < 0) return -1;

    pid_t pid = fork();
    if (pid < 0) {
        close(pipefd[0]); close(pipefd[1]);
        return -1;
    }

    if (pid == 0) {
        /* child: redirect stdout to pipe write end, exec screencap */
        close(pipefd[0]);
        dup2(pipefd[1], STDOUT_FILENO);
        close(pipefd[1]);
        execl("/system/bin/screencap", "screencap", NULL);
        _exit(1);
    }

    /* parent: read from pipe read end, send to client */
    close(pipefd[1]);
    char buf[65536];
    ssize_t n;
    while ((n = read(pipefd[0], buf, sizeof(buf))) > 0) {
        ssize_t total = 0;
        while (total < n) {
            ssize_t w = send(client_fd, buf + total, n - total, 0);
            if (w <= 0) { close(pipefd[0]); waitpid(pid, NULL, 0); return -1; }
            total += w;
        }
    }
    close(pipefd[0]);
    waitpid(pid, NULL, 0);
    return 0;
}

static void handle_client(int client_fd) {
    char cmd[64];
    memset(cmd, 0, sizeof(cmd));
    ssize_t n = recv(client_fd, cmd, sizeof(cmd) - 1, 0);
    if (n <= 0) return;

    /* strip whitespace */
    while (n > 0 && (cmd[n-1] == '\n' || cmd[n-1] == '\r' || cmd[n-1] == ' '))
        cmd[--n] = '\0';

    if (strcmp(cmd, "ping") == 0) {
        send(client_fd, "pong\n", 5, 0);
    } else if (strcmp(cmd, "shot") == 0) {
        send_screenshot(client_fd);
    }
}

int main(int argc, char *argv[]) {
    int port = (argc > 1) ? atoi(argv[1]) : 19000;

    /* ignore SIGCHLD to auto-reap children */
    signal(SIGCHLD, SIG_IGN);

    int srv = socket(AF_INET, SOCK_STREAM, 0);
    if (srv < 0) { perror("socket"); return 1; }

    int opt = 1;
    setsockopt(srv, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt));

    struct sockaddr_in addr;
    memset(&addr, 0, sizeof(addr));
    addr.sin_family = AF_INET;
    addr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
    addr.sin_port = htons(port);

    if (bind(srv, (struct sockaddr*)&addr, sizeof(addr)) < 0) {
        perror("bind"); return 1;
    }
    if (listen(srv, 4) < 0) {
        perror("listen"); return 1;
    }

    fprintf(stdout, "Screenshot daemon listening on port %d\n", port);
    fflush(stdout);

    while (1) {
        struct sockaddr_in cli;
        socklen_t cli_len = sizeof(cli);
        int client = accept(srv, (struct sockaddr*)&cli, &cli_len);
        if (client < 0) continue;
        handle_client(client);
        close(client);
    }
    return 0;
}
